platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java

```diff
@@ -52,16 +55,28 @@
 
   private String metadata;
 
+  private Instant created;
+
+  private Instant validUntil;
+
+  private Duration cacheDuration;
+
   private AtomicReference<Map<String, EntityDescriptor>> entryDescriptions =
       new AtomicReference<>();
 
   private String singleLogoutBinding;
 
   private String singleLogoutLocation;
 
-  public void setMetadata(String metadata) {
+  public synchronized void setMetadata(String metadata) {
```

Comment:
✏️ There's quite a bit of added complexity and the potential for a non-trivial performance hit under load with the addition of the synchronization. I quickly hacked together an alternative I'd like you to consider. It needs to be reviewed and tested thoroughly and I'm not quite sure about the implementations of the `isMetadataExpired` and `isMetadataValid` methods in my version, but it should otherwise be equivalent and a bit more performant.

Take a look here - https://gist.github.com/coyotesqrl/89cd9ba8b231df4485b2e46d6886c03c

Reply:
I had thought about wrapping together the entity description and the other state variables. I like it better than slathering `synchronized` keywords all over the class.

It seems odd that we store a map of EntityDescriptors, but only reference the first one. What do the others represent, I wonder?  Under what circumstances would we reference them?

Reply:
I don't recall. @stustison could you weigh in?